docs/src/design/multiset.md

```diff
@@ -1,56 +1,158 @@
+
+The consistency of a virtual table can be proved with a single trace column $p$, which keeps a running product of rows that were inserted into and deleted from the table. This is done by reducing each row to a single value, multiplying the value into $p$ when the row is inserted, and dividing the value out of $p$ when the row is removed. Thus, at any step of the computation, $p$​ will contain a product of all rows currently in the table. 
+
+The initial value of $p$​ is set to 1. Thus, if the table is empty by the time Miden VM finishes executing a program (we added and then removed exactly the same set of rows), the final value of $p$​ will also be equal to 1. The initial and final values are enforced via boundary constraints.
+
+### Computing a virtual table's trace column
+
+To compute a product of rows, we'll first need to reduce each row to a single value. This can be done as follows.
 
-To compute a product of rows, we'll first need to reduce each row to a single value. This will be done as follows. Assume the verifier sends a set of random values $\alpha_0$, $\alpha_1$ etc. to the prover after the prover commits to the execution trace of the program. The prover reduces row $i$ in the table to a single value $r_i$ as:
+Let $t_0, t_1, t_2, ...$ be columns in the virtual table, and assume the verifier sends a set of random values $\alpha_0$, $\alpha_1, ...$ to the prover after the prover commits to the execution trace of the program.
+
+The prover reduces row $i$ in the table to a single value $r_i$ as:
 
 $$
-r_i = \alpha_0 + \alpha_1 \cdot t_{0, i} + \alpha_2 \cdot t_{1, i} + \alpha_3 \cdot t_{2, i}
+r_i = \alpha_0 + \alpha_1 \cdot t_{0, i} + \alpha_2 \cdot t_{1, i} + \alpha_3 \cdot t_{2, i} + ...
 $$
 
-Then, when row $i$ is added to the table, we'll update the value in the $p_1$ column like so:
+Then, when row $i$ is added to the table, we'll update the value in the $p$ column like so:
 
 $$
-p_1' = p_1 \cdot r_i
+p' = p \cdot r_i
 $$
 
-where, $p_1$ denotes the value of column $p_1$ at the current step and $p_1'$ denotes the value of column $p_1$ at the next step.
+Analogously, when row $i$ is removed from the table, we'll update the value in column $p$ like so:
+
+$$
+p' = \frac{p}{r_i}
+$$
 
-Analogously, when row $i$ is removed from the table, we'll update the value in column $p_1$ like so:
+In this general approach, a value was included in $p$ for each column $t_j$ in the virtual table. However, the row reduction could also combine multiple columns into a single value for inclusion. For example, if $T$ is a 4-element value stored in columns $t_2, t_3, t_4, t_5$, row $i$ could be reduced to $r_i$ as:
 
 $$
-p_1' = \frac{p_1}{r_i}
+r_i = \alpha_0 + \alpha_1 \cdot t_{0, i} + \alpha_2 \cdot t_{1, i} + \alpha_3 \cdot T_i + \alpha_4 \cdot t_6 + ...
 $$
```

Comment:
I am not sure I followed this section. We still need to multiply each element of $T$ by "its own" alpha. Is there a typo here somewhere?

Reply:
Oops, no, it's just an error - I think I was just looking at this too long and got turned around by another part of the docs. I'll cut it in the edit PR.